# BÀI TẬP: TITANIC
**Nguồn:** kaggle.com/c/titanic (891 dòng)


In [1]:
import pandas as pd, numpy as np, matplotlib.pyplot as plt, seaborn as sns
from pathlib import Path
from scipy import stats

sns.set_style('whitegrid')

csv_path = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'

df = pd.read_csv(csv_path)
print('Loaded from:', csv_path)
df.head()

Loaded from: https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv


,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


---
# PHẦN A — DATA PROFILING
## A.1. Data size, column names, data types

In [2]:
# TODO
df.shape
df.columns.to_list()
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    str    
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         891 non-null    float64
 7   embarked     889 non-null    str    
 8   class        891 non-null    str    
 9   who          891 non-null    str    
 10  adult_male   891 non-null    bool   
 11  deck         203 non-null    str    
 12  embark_town  889 non-null    str    
 13  alive        891 non-null    str    
 14  alone        891 non-null    bool   
dtypes: bool(2), float64(2), int64(4), str(7)
memory usage: 92.4 KB


## A.2. Missing values & Duplicate data

In [3]:
# TODO
print(df.isnull().sum())
print(df.duplicated().sum())

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64
107


## A.3. Invalid values

In [4]:
df.describe()

,survived,pclass,age,sibsp,parch,fare
count,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


## A.4. Create a new column
Tạo cột `family_size` = sibsp + parch + 1.

In [5]:
# TODO
df['family_size'] = df['sibsp'] + df['parch'] + 1
df.head()

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone,family_size
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False,2
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False,2
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True,1
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False,2
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True,1


---
# PHẦN B — DESCRIPTIVE STATISTICS
## Group 1 — Central Tendency

In [6]:
df.mean(numeric_only=True)
df.median(numeric_only=True)
df.mode(numeric_only=True).iloc[0]

survived          0
pclass            3
age            24.0
sibsp             0
parch             0
fare           8.05
adult_male     True
alone          True
family_size       1
Name: 0, dtype: object

## Group 2 — Dispersion

In [10]:
df.std(numeric_only=True)
df.var(numeric_only=True)

survived          0.236772
pclass            0.699015
age             211.019125
sibsp             1.216043
parch             0.649728
fare           2469.436846
adult_male        0.239723
alone             0.239723
family_size       2.603248
dtype: float64

## Group 3 — Location and Shape

In [11]:
num_cols = df.select_dtypes(include=np.number)

print(num_cols.quantile([0.25, 0.5, 0.75]))
print(num_cols.skew())
print(num_cols.kurtosis())

      survived  pclass     age  sibsp  parch     fare  family_size
0.25       0.0     2.0  20.125    0.0    0.0   7.9104          1.0
0.50       0.0     3.0  28.000    0.0    0.0  14.4542          1.0
0.75       1.0     3.0  38.000    1.0    0.0  31.0000          2.0
survived       0.478523
pclass        -0.630548
age            0.389108
sibsp          3.695352
parch          2.749117
fare           4.787317
family_size    2.727441
dtype: float64
survived       -1.775005
pclass         -1.280015
age             0.178274
sibsp          17.880420
parch           9.778125
fare           33.398141
family_size     9.159666
dtype: float64


---
# PHẦN C — DEFINE THE QUESTION

## Câu hỏi 1: Hạng vé nào có tỷ lệ sống sót cao nhất, chênh lệch bao nhiêu so với hạng thấp nhất?

In [14]:
p1 = df.groupby('pclass')['survived'].mean()
print(p1, "\nChênh lệch:", p1.max() - p1.min())

pclass
1    0.629630
2    0.472826
3    0.242363
Name: survived, dtype: float64 
Chênh lệch: 0.3872671041713812


## Câu hỏi 2: Giới tính hay hạng vé ảnh hưởng đến sống sót mạnh hơn?

In [15]:
print("Sex diff:", df.groupby('sex')['survived'].mean().pipe(lambda x: x.max() - x.min()))
print("Pclass diff:", df.groupby('pclass')['survived'].mean().pipe(lambda x: x.max() - x.min()))

Sex diff: 0.5531300709799203
Pclass diff: 0.3872671041713812


## Câu hỏi 3: Vé đắt hơn có thực sự sống sót cao hơn không?

In [16]:
df.groupby('survived')['fare'].mean()

survived
0    22.117887
1    48.395408
Name: fare, dtype: float64

## Câu hỏi 4: Gia đình đông người có ảnh hưởng đến khả năng sống sót không?

In [17]:
df['family_size'] = df['sibsp'] + df['parch'] + 1
df.groupby('family_size')['survived'].mean()

family_size
1     0.303538
2     0.552795
3     0.578431
4     0.724138
5     0.200000
6     0.136364
7     0.333333
8     0.000000
11    0.000000
Name: survived, dtype: float64

## Câu hỏi 5: Cảng lên tàu (embark_town) nào có tỷ lệ sống sót cao nhất?

In [18]:
# TODO
print(df.groupby('embark_town')['survived'].mean().sort_values(ascending=False))

embark_town
Cherbourg      0.553571
Queenstown     0.389610
Southampton    0.336957
Name: survived, dtype: float64


## Câu hỏi 6 (Tổng hợp) — Viết insight tổng hợp
Dựa trên Phần A, B, C, viết 4-5 câu insight tổng thể về dữ liệu Titanic.

*(Viết insight của bạn vào đây...)*

1. Phụ nữ có tỷ lệ sống sót cao vượt trội so với nam giới (~74% so với ~19%).
2. Hạng vé càng cao (vé đắt tiền) thì khả năng sống sót càng lớn (Hạng 1 cao nhất ~63%).
3. Đi cùng gia đình nhỏ (2-4 người) có tỷ lệ sống sót cao hơn đi một mình hoặc gia đình quá đông.
4. Cảng Cherbourg có tỷ lệ cứu nạn cao nhất do tập trung nhiều khách hạng nhất.